sdk/iot/core/src/az_iot_hub_client.c

```diff
@@ -1,12 +1,137 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
+#include <stdint.h>
+
 #include "az_iot_hub_client.h"
+#include <az_precondition_internal.h>
+#include <az_result.h>
 #include <az_span.h>
 
 #include <_az_cfg.h>
 
+static const uint8_t hub_client_forward_slash = '/';
+static const uint8_t hub_client_param_separator = '&';
+static const uint8_t hub_client_null_terminate = '\0';
+
+static const az_span hub_client_api_version = AZ_SPAN_LITERAL_FROM_STR("?api-version=2018-06-30");
+
 AZ_NODISCARD az_iot_hub_client_options az_iot_hub_client_options_default()
 {
   return (az_iot_hub_client_options){ .module_id = AZ_SPAN_NULL, .user_agent = AZ_SPAN_NULL };
 }
+
+AZ_NODISCARD az_result az_iot_hub_client_init(
+    az_iot_hub_client* client,
+    az_span iot_hub_hostname,
+    az_span device_id,
+    az_iot_hub_client_options const* options)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(iot_hub_hostname, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(device_id, 1, false);
+  AZ_PRECONDITION_NOT_NULL(options);
+
+  client->_internal.iot_hub_hostname = iot_hub_hostname;
+  client->_internal.device_id = device_id;
+  memcpy(&client->_internal.options, options, sizeof(az_iot_hub_client_options));
+
+  return AZ_OK;
+}
+
+//[Format without module id] {iothubhostname}/{device_id}/?api-version=2018-06-30
+//[Format with module id] {iothubhostname}/{device_id}/{module_id}/?api-version=2018-06-30
+AZ_NODISCARD az_result az_iot_hub_client_user_name_get(
+    az_iot_hub_client const* client,
+    az_span mqtt_user_name,
+    az_span* out_mqtt_user_name)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
```

Comment:
why are we are not allowing NULL out? If user does not want out span he should be allowed, atleast that is how sdk-core span related apis are designed. Am I missing anything in this API?

Reply:
same comment for all API impl

---
In reply to: [392637338](https://github.com/Azure/azure-sdk-for-c/pull/455#discussion_r392637338) [](ancestors = 392637338)

Reply:
I'm not sure which API's use that dynamic but ones like this
https://github.com/Azure/azure-sdk-for-c/blob/34c8b78dd70d07c93e8fb63b3e865351da223a5e/sdk/core/core/src/az_span.c#L160-L165
or
https://github.com/Azure/azure-sdk-for-c/blob/34c8b78dd70d07c93e8fb63b3e865351da223a5e/sdk/core/core/src/az_span.c#L223-L225
require the `out_xxx` param to not be `NULL`. If you see some that don't require that maybe we can talk to the Azure SDK for C team to rectify inconsistencies.

Reply:
talked with Jeffrey and the output span should be valid. so we will check for `NULL`.